Adding new managed WordPress hosts to be identified in class-functions.php.

oikeme opened the pull request:
Adding more hosting providers to get_hosting_provider() function to be able to identify managed WordPress hosts.
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Fixes N/A

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Adding options to identify new hosts specifically Managed WordPress hosts.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* New Feature pbtFPC-1M-p2

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Unit Tests: phpunit --filter=WP_Test_Jetpack_Sync_Functions
#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Add more managed WordPress hosts to hosting provider detection.

oikeme committed:
Update class-functions

Adding more hosting providers to get_hosting_provider() function to be able to identify managed WordPress hosts.

oikeme committed:
Update class-functions.php

Update get_hosting_provider() function to use an array of hosting provider constants to check the hosting provider

ebinnion commented on packages/sync/src/class-functions.php:
```diff
@@ -161,15 +161,21 @@ public static function get_post_type_features() {
+			'IS_PRESSABLE'         => 'pressable',
+		);
+
+		foreach ( $hosting_provider_constants as $constant => $constant_value ) {
```
Instead of `$constant_value`, perhaps we can use `$hosting_provider`?

oikeme replied:
Makes more sense. I've updated this

oikeme committed:
Update class-functions.php

Adding in secondary check for class_exists for gd-managed-wp

oikeme committed:
Update class-functions.php

Update variable name when returning the constant value to $hosting_provider

oikeme committed:
Update hosting provider check for managed WP and adding tests

Update checking for managed WP hosts and adding tests for get_hosting_provider() function check.

tyxla commented on tests/php/sync/test_class.jetpack-sync-callables.php:
```diff
@@ -1053,6 +1053,50 @@ public function test_sync_callable_recursive_gets_checksum() {
+	 */
+	public function test_get_hosting_provider_callable() {
+		Constants::set_constant( 'GD_SYSTEM_PLUGIN_DIR', 'set' );
+		if ( Constants::is_defined( 'GD_SYSTEM_PLUGIN_DIR' ) || class_exists( '\\WPaaS\\Plugin' ) ) {
```
Drive-by comment: it feels a bit odd to have those checks in tests. I'd expect that the `::is_defined` check is not necessary because we just set the constant on the previous line; and instead of the `class_exists` check I'd suggest defining the class or even better - wrapping the `class_exists` check in another method that could be mocked for testing purposes.

oikeme replied:
@tyxla thanks for the suggestion on this! It led me to make a couple of other changes to including splitting out the different types of checks and updating the tests.

ebinnion commented:
This is looking pretty good to me. I agree with @tyxla's feedback on the the tests.

Besides that, I wouldn't worry much about the Code Climate issue. It seems like it doesn't like the length of the method, but that may be difficult to fix and frankly, the method will continue to grow as we we add hosts in the future. I would suggest updating/removing the functional testing instructions since you've added test coverage.

oikeme committed:
Update class-functions.php

Splitting out hosting provider detection in to detection by specific types in separate functions which are now called by the main get_hosting_provider() function

oikeme committed:
Update Tests for get_hosting_provider

Updating class functions to be non static and adding test for get_hosting_provider() for each method of getting a hosting provider

oikeme committed:
Update unit tests and fix get_hosting_provider() return

Updating the unit tests for get_hosting_provider() to be separate tests for each method that function calls and a test for unknown hosting providers from the get_hosting_provider() callable

ebinnion reviewed (dismissed):
This LGTM. But, I'd like to get one other set of eyes on it as well.

oikeme commented:
@tyxla I've made some changes to this patch since your suggestion which was great by the way. Do you mind giving it a look over again? Thanks!

tyxla commented on packages/sync/src/class-functions.php:
```diff
@@ -161,22 +161,90 @@ public static function get_post_type_features() {
+		$hosting_provider = false;
+
+		switch ( true ) {
+			case ( Constants::is_defined( 'GD_SYSTEM_PLUGIN_DIR' ) ):
```
I think that it would be a bit more concise to change this `switch` to either:

* An array map of constant => hosting provider, with a loop and a return if there's a match
* An `if/else`

mostly because it isn't a real syntactic `switch` because it doesn't check for a variable.

oikeme replied:
Good call. I've changed this. Updated previously due to not using the Constants class but this is definitely more concise with the loop.

tyxla commented on packages/sync/src/class-functions.php:
```diff
@@ -161,22 +161,90 @@ public static function get_post_type_features() {
+		$hosting_provider = false;
+
+		switch ( true ) {
+			case ( class_exists( '\\WPaaS\\Plugin' ) ):
```
The same comment from above applies for this `switch`, too.

tyxla commented on packages/sync/src/class-functions.php:
```diff
@@ -161,22 +161,90 @@ public static function get_post_type_features() {
+	public function get_hosting_provider_by_known_function() {
+		$hosting_provider = false;
+
+		switch ( true ) {
```
The same comment from above applies for this `switch`, too.

tyxla reviewed (dismissed):
This looks good from my perspective 👍 

I'd only suggest against the way that `switch(true)` is used, but I don't have super strong feelings if y'all prefer it to be that way.

oikeme committed:
Update class-functions.php

Update function get_hosting_provider_by_known_constant() to loop through known hosting provider constants in place of a switch

ebinnion commented on packages/sync/src/class-functions.php:
```diff
@@ -161,22 +161,83 @@ public static function get_post_type_features() {
+
+		foreach ( $hosting_provider_constants as $constant => $constant_value ) {
+			if ( Constants::is_defined( $constant ) ) {
+				if ( 'VIP_GO_ENV' === $constant && false === Constants::get_constant( 'VIP_GO_ENV' ) ) {
```
I wonder if it makes sense to apply this `false` check to all of the constants? Maybe something like:

```
if ( Constants::is_defined( $constant ) && false !== Constants::get_constant( 'VIP_GO_ENV' )  ) {
  return $constant_value;
}
```

This isn't a blocker, but it could minimize the need for us to change this in the future.

oikeme replied:
Wasn't sure on this as this logic is from before the update and specific to vip-go but it's a good call. I've updated this for everything.

tyxla replied:
Seems like this still needs to be updated though as it's `VIP_GO_ENV`-specific?

ebinnion commented on packages/sync/src/class-functions.php:
```diff
@@ -161,22 +161,83 @@ public static function get_post_type_features() {
+		$hosting_provider = false;
+
+		switch ( true ) {
+			case ( function_exists( 'is_wpe' ) || function_exists( 'is_wpe_snapshot' ) ):
```
Personally, I'd prefer this as two separate case statements.


```
case function_exists( 'is_wpe' ):
case  function_exists( 'is_wpe_snapshot' ) ):
  $hosting_provider = 'wpe';
  break;
```

But, that's just a personal preference. We don't need to change it here.

oikeme replied:
I missed this in the prior commit but switched this to a loop instead

ebinnion reviewed (approved):
I left a couple of comments, but they're not blockers. I like the change to use an array instead of a case for checking constants. It seems to have simplified the code quite a bit. 

I think it's fine to not make the same change to the other methods for now at least. Instead, we can make that change if/when necessary, as we add more functions and classes to check.

oikeme commented:
@tyxla I had a comment from you in an email earlier today about an issue I think may have been resolved. Can you confirm?

jeherve reviewed (approved):
This looks good to me! 👍

tyxla commented:
I can see some of the comments were marked as resolved but not addressed, but they were also not blockers. So feel free to 🚢 if it's all intended 👍
